pyvkfft/opencl.py

```diff
@@ -155,20 +155,24 @@ def _make_config(self):
                                          int(self.use_lut), int(self.keepShaderCode),
                                          n_batch, skipx, skipy, skipz)
 
-    def fft(self, src: cla.Array, dest: cla.Array = None):
+    def fft(self, src: cla.Array, dest: cla.Array = None, queue: cl.CommandQueue = None):
```

Comment:
Note that the array also carries a queue.

Reply:
Which can be None right? Any suggestions on what to do?

Reply:
Yes, it can be `None`, but I think it's reasonable to expect an array passed in here to have a valid queue.

If you were designing this API from scratch, I think the following fallback order would make sense:

- `queue` (if passed)
- `src.queue` and `dest.queue` (if given, must be the same)
- I would not consider `self.queue`.

For backward compatibility, I would warn if `self.queue` disagrees with queue resulting from the above and say that `self.queue` is being used *for now*, but that that behavior is deprecated, with a certain fuse.

Reply:
Note that the 'simple' fft API (`pyvkfft.fft.fftn`, `pyvkfft.fft.ifftn`...) already allows to transform just by passing an array - its queue will be used as default, and a different one can be supplied. This already allows to use different queues for different transforms.

> If you were designing this API from scratch, I think the following fallback order would make sense:
> 
> * `queue` (if passed)
> * `src.queue` and `dest.queue` (if given, must be the same)
> * I would not consider self.queue.
>
> For backward compatibility, I would warn if `self.queue` disagrees with queue resulting from the above and say that `self.queue` is being used for now, but that that behavior is deprecated, with a certain fuse.

There can be some strange results when different queues are used, e.g. when using a different queue from the array's, the get() method will not wait for the fft to be finished.

I would still keep `VkFFTApp`'s `self.queue` as the default (assuming it's a deliberate choice from the developer), but a warning about when different queues are used will be useful.

Reply:
The queue choice is indeed debatable.. Preferring the array queues could also make sense.. It really depends on the use case.

The main use case is probably not a single array which is transformed with different queues at different times, but rather different arrays which are transformed by the same `VkFFTApp`. In which case using the array queues would indeed be easier...